c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -328,10 +310,106 @@ cudaError_t cudaMallocMaybeCapturing(void** p, size_t size) {
 #endif
 }
 
-} // namespace
+} // anonymous namespace
+} // namespace Native
+
+// Backend static initialization.
+#define DECLARE_BACKEND_INTERFACE(RET, FUNC, ARGS) \
+  RET FUNC ARGS;
+
+// Not called directly by clients.
+namespace Native {
+FORALL_ALLOCATOR_INTERFACE(DECLARE_BACKEND_INTERFACE)
+}
+
+// Not called directly by clients.
+namespace CudaMallocAsync {
+FORALL_ALLOCATOR_INTERFACE(DECLARE_BACKEND_INTERFACE)
+}
+
+#undef DECLARE_BACKEND_INTERFACE
+
+#define DEFINE_CHOSEN(RET, FUNC, ARGS) \
+  RET (*FUNC) ARGS = 0;
 
+namespace Chosen {
+  FORALL_ALLOCATOR_INTERFACE(DEFINE_CHOSEN);
+} // namespace Chosen
+
+#define INITIALIZE_NATIVE(RET, FUNC, ARGS) \
+  Chosen::FUNC = Native::FUNC;
+
+#define INITIALIZE_CUDAMALLOCASYNC(RET, FUNC, ARGS) \
+  Chosen::FUNC = CudaMallocAsync::FUNC;
+
+struct BackendStaticInitializer {
+  AllocatorBackend backend;
+
+  // Parses env for backend at load time, duplicating some logic from CachingAllocatorConfig.
+  // CachingAllocatorConfig double-checks it later (at runtime).
+  // Defers verbose exceptions and error checks, including Cuda version checks,
+  // to CachingAllocatorConfig's runtime doublecheck.
+  // If this works, maybe we should move all of CachingAllocatorConfig here?
+  AllocatorBackend parseEnvForBackend() {
+    const char* val = getenv("PYTORCH_CUDA_ALLOC_CONF");
```

Comment:
@swolchok how does parsing things out of envvars in static initializers make you feel

Reply:
seems fine as long as you don't mind failing to respect changes to the environment via putenv() etc. I don't think I have ever minded.